snakemake/subworkflow.py

```diff
@@ -0,0 +1,66 @@
+import os
+
+from snakemake.logging import logger
+from snakemake.exceptions import WorkflowError
+
+
+def subworkflow(
+    path,
+    basedir,
+    input,
+    output,
+    params,
+    wildcards,
+    threads,
+    resources,
+    log,
+    config,
+    rulename,
+    conda_env,
+    container_img,
+    singularity_args,
+    env_modules,
+    bench_record,
+    jobid,
+    bench_iteration,
+    cleanup_scripts,
+    shadow_dir,
+    workflow,
+):
+    # extract (and normalize) parameters
+    basedir = os.path.normpath(basedir)
+    sub_workdir = os.path.normpath(params.get("workdir", f"subworkflow_{rulename}"))
+    sub_configfile = params.get("configfile", "")
+
+    # the sub-workflow needs to run in a different directory than the
+    # main-workflow. This is necessary due to Snakemake's locking mechanism
+    if basedir == sub_workdir:
+        raise WorkflowError(
+            "Sub-workflow workdir must be different from main-workflow workdir"
+        )
+
+    # all output files must use the sub-workflow's workdir as their prefix
+    # otherwise file dependencies cannot be easily handled
```

Comment:
I don't get why this is the case?  Why is this a requirement?  

I can see many cases where I'll have many sub-workflows all outputting to the same output directory.

Reply:
I'll discuss this further in a reply in the main thread.